wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceUpgrader.java

```diff
@@ -0,0 +1,140 @@
+    });
+
+    for (Workspace workspace : workspaces) {
+      List<Row> sample = Collections.emptyList();
+      try {
+        sample = DirectivesHandler.fromWorkspace(workspace);
+      } catch (Exception e) {
+        // this should not happen, but guard here to avoid failing the entire upgrade process
+        LOG.warn("Error retrieving the sample for workspace {}", workspace.getName(), e);
+      }
+      List<String> directives = workspace.getRequest() == null ? Collections.emptyList() :
+                                  workspace.getRequest().getRecipe().getDirectives();
+      WorkspaceId workspaceId = new WorkspaceId(namespace, workspace.getNamespacedId().getId());
+
+      long now = System.currentTimeMillis();
+      io.cdap.wrangler.proto.workspace.v2.Workspace.Builder ws =
+        io.cdap.wrangler.proto.workspace.v2.Workspace
+          .builder(workspace.getName(), workspace.getNamespacedId().getId()).setDirectives(directives)
+          .setCreatedTimeMillis(now).setUpdatedTimeMillis(now);
+
+      String connectorName = SpecificationUpgradeUtils.getConnectorName(
+        workspace.getProperties().get(PropertyIds.CONNECTION_TYPE).toLowerCase());
+      String path = SpecificationUpgradeUtils.getPath(workspace);
+      // if the connection type cannot be upgraded, create a workspace with empty sample spec, so there will be
+      // no sources created when converting to pipeline
+      if (path == null) {
+        wsStore.saveWorkspace(workspaceId, new WorkspaceDetail(ws.build(), sample));
+        continue;
+      }
+
+      String connection = workspace.getProperties().get(PropertyIds.CONNECTION_ID);
+      try {
+        ConnectorDetail detail =
+          discoverer.getSpecification(namespace.getName(), connection,
+                                      new SpecGenerationRequest(path, Collections.emptyMap()));
+
+        SampleSpec spec = new SampleSpec(
+          connection, connectorName, path,
+          detail.getRelatedPlugins().stream().map(plugin -> {
+            ArtifactSelectorConfig artifact = plugin.getArtifact();
+            Plugin pluginSpec = new Plugin(
+              plugin.getName(), plugin.getType(), plugin.getProperties(),
+              new Artifact(artifact.getName(), artifact.getVersion(), artifact.getScope()));
+            return new StageSpec(plugin.getSchema(), pluginSpec);
+          }).collect(Collectors.toSet()));
+        wsStore.saveWorkspace(workspaceId, new WorkspaceDetail(ws.setSampleSpec(spec).build(), sample));
+      } catch (NotFoundException e) {
+        LOG.warn("Connection {} related to workspace {} does not exist. " +
+                   "The workspace will be upgraded without that information", connection, workspace.getName());
+        wsStore.saveWorkspace(workspaceId, new WorkspaceDetail(ws.build(), sample));
+      } catch (Exception e) {
+        LOG.warn("Unable to get the spec from connection {} for workspace {}. " +
```

Comment:
when does this happen? Is it a transient exception? Does it make more sense to retry for some time and then give up?

upgrading this workspace without the spec would put it in a broken state that could never be fixed right? Seems better not to upgrade it at all rather than do that.

Reply:
The get spec call in discoverer already have retry logic, it will only throw exception when a non-retryable exception happens or retry timed out. 
This can happen for database related connectors where the url does not include the database name, I think we should still upgrade them to let them show up in the new UX

Reply:
Actually if we are able to create the workspace for db relations, it is quite possible we are able to get spec back, changed this to not upgrade and just leave an error message